src/snowflake/snowpark/_internal/data_source/drivers/psycopg2_driver.py

```diff
@@ -0,0 +1,362 @@
+        # the default behavior of psycopg2 is to get Interval as datetime.timedelta
+        # https://other-docs.snowflake.com/en/connectors/postgres6/view-data#postgresql-to-snowflake-data-type-mapping
+        from psycopg2.extensions import new_type, register_type
+
+        SNOWPARK_INTERVAL_STR = new_type(
+            (Psycopg2TypeCode.INTERVALOID.value,),
+            "SNOWPARK_INTERVAL_STR",
+            lambda data, cursor: data,
+        )
+        register_type(SNOWPARK_INTERVAL_STR, conn)
+
+        # by default psycopg2 returns binary data as memoryview
+        # to avoid using pandas to convert memoryview to bytes, we use the following native psycopg2 type conversion
+        # psycopg2.extensions.new_type() only works for text format data, it returns bytes as hex string
+        # we reconstruct the bytes from hex string
+        SNOWPARK_BYTE = new_type(
+            (Psycopg2TypeCode.BYTEAOID.value,),
+            "SNOWPARK_BYTE_BYTES",
+            lambda data, cursor: bytes.fromhex(data[2:])
+            if data is not None
+            else None,  # [2:] to skip the '\\x' prefix
+        )
+        register_type(SNOWPARK_BYTE, conn)
+
+        if query_timeout:
+            # https://www.postgresql.org/docs/current/runtime-config-client.html#GUC-STATEMENT-TIMEOUT
+            # postgres default uses milliseconds
+            conn.cursor().execute(f"SET STATEMENT_TIMEOUT = {query_timeout * 1000}")
+        return conn
+
+    def udtf_class_builder(
+        self, fetch_size: int = 1000, schema: StructType = None
+    ) -> type:
+        create_connection = self.create_connection
+
+        # TODO: SNOW-2101485 use class method to prepare connection
+        # ideally we should use the same function as prepare_connection
+        # however, since we introduce new module for new driver support and initially the new module is not available in the backend
+        # so if registering UDTF which uses the class method, cloudpickle will pickle the class method along with
+        # the new module -- this leads to not being able to find the new module when unpickling on the backend.
+        # once the new module is available in the backend, we can use the class method.
+        def prepare_connection_in_udtf(
+            conn: "Connection",
+            query_timeout: int = 0,
+        ) -> "Connection":
+            # The following is to align with Snowflake Connector behavior which get Interval as string
+            # the default behavior of psycopg2 is to get Interval as datetime.timedelta
+            # https://other-docs.snowflake.com/en/connectors/postgres6/view-data#postgresql-to-snowflake-data-type-mapping
+            from psycopg2.extensions import new_type, register_type
+
+            # we do not use Psycopg2TypeCode.INTERVALOID.value because UTDF pickles the psycopg2_driver module
+            # unpickling in the UDTF would results in module not found error if package not available in the backend
```

Comment:
does this means we need to remove this after package is available?

Reply:
yes, we can remove the local function once the new module arrived backend